packages/flutter/lib/src/painting/_network_image_io.dart

```diff
@@ -57,63 +58,121 @@ class NetworkImage extends image_provider.ImageProvider<image_provider.NetworkIm
-  // in our call to [consolidateHttpClientResponseBytes].
-  static final HttpClient _sharedHttpClient = HttpClient()..autoUncompress = false;
-
-  static HttpClient get _httpClient {
-    HttpClient client = _sharedHttpClient;
-    assert(() {
-      if (debugNetworkImageHttpClientProvider != null)
-        client = debugNetworkImageHttpClientProvider();
-      return true;
-    }());
-    return client;
-  }
+  static Future<Isolate> _pendingLoader;
+  static List<_PendingLoadRequest> _pendingLoadRequests;
+  static SendPort _requestPort;
 
-  Future<ui.Codec> _loadAsync(
+  Future<ui.Codec> _loadAsyncOnDesignatedIsolate(
     NetworkImage key,
     StreamController<ImageChunkEvent> chunkEvents,
   ) async {
     try {
       assert(key == this);
 
       final Uri resolved = Uri.base.resolve(key.url);
-      final HttpClientRequest request = await _httpClient.getUrl(resolved);
-      headers?.forEach((String name, String value) {
-        request.headers.add(name, value);
+
+      final Completer<TransferableTypedData> completer = Completer<TransferableTypedData>();
+      final RawReceivePort port = RawReceivePort((dynamic message) {
+        if (message is TransferableTypedData) {
+          completer.complete(message);
+        } else if (message is ImageChunkEvent) {
+          chunkEvents.add(message);
+        } else {
+          completer.completeError(message);
+        }
       });
-      final HttpClientResponse response = await request.close();
-      if (response.statusCode != HttpStatus.ok)
-        throw Exception('HTTP request failed, statusCode: ${response?.statusCode}, $resolved');
-
-      final Uint8List bytes = await consolidateHttpClientResponseBytes(
-        response,
-        onBytesReceived: (int cumulative, int total) {
-          chunkEvents.add(ImageChunkEvent(
-            cumulativeBytesLoaded: cumulative,
-            expectedTotalBytes: total,
-          ));
-        },
-      );
-      if (bytes.lengthInBytes == 0)
+
+      // This will keep reference to [debugNetworkImageHttpClientProvider] tree-shaken
+      // out of release build.
+      HttpClientProvider httpClientProvider;
+      assert(() { httpClientProvider = debugNetworkImageHttpClientProvider; return true; }());
+
+      final _DownloadRequest downloadRequest = _DownloadRequest(port.sendPort,
+          resolved, headers, httpClientProvider);
+      if (_requestPort != null) {
+        // If worker isolate is properly set up([_requestPort] is holding
+        // initialized [SendPort], then just send download request down to it.
+        _requestPort.send(downloadRequest);
+      } else {
+        if (_pendingLoader == null) {
+          // If worker isolate creation was not started, start creation now.
+          _pendingLoadRequests = <_PendingLoadRequest>[];
+
+          _pendingLoader = _setupIsolate()..then((Isolate isolate) {
+              isolate.addErrorListener(RawReceivePort(
+                  (List<dynamic> errorAndStacktrace) {
```

Comment:
Yep, done.

Reply:
Actually, no. It has to be defined as List<dynamic> - otherwise you will get `type 'List<dynamic>' is not a subtype of type 'List<String>' of 'errorAndStackTrace'` errors when this handler is triggered.